Simplify Container Status Patching

kflansburg opened the pull request:
Closes #410

kflansburg commented on crates/kubelet/src/container/status.rs:
```diff
@@ -82,3 +85,90 @@ impl Status {
 }
+
+/// Patch a single container's status
+pub async fn patch_container_status(
```
This relies on container statuses matching the containers in `Pod.containers` and `Pod.init_containers`, in the same order. I believe that this would be best practice.

kflansburg commented on crates/kubelet/src/pod/mod.rs:
```diff
@@ -164,16 +166,11 @@ impl Pod {
-        let app_container_keys = app_containers
-            .iter()
-            .map(|c| ContainerKey::App(c.name().to_owned()));
+    pub fn all_containers(&self) -> Vec<Container> {
```
This seemed inconsistent with the other two methods `containers` and `init_containers`.

itowlson replied:
I suspect at the time it was written it was used primarily for getting keys/IDs rather than the full container objects. But it looks like now the only places it is used would work just as well with the full container objects.  There is also a place in `store/mod.rs` that explicitly constructs its own `all_containers` collection which this modified method could supplant.

kflansburg commented on crates/kubelet/src/pod/status.rs:
```diff
@@ -1,8 +1,28 @@
 
-use crate::container::{ContainerMap, Status as ContainerStatus};
+/// Initialize Pod status.
+pub fn make_registered_status(pod: &Pod) -> anyhow::Result<serde_json::Value> {
```
This initializes Pod status to include containers in the correct order as expected by `patch_container_status`. It is kind of required, so I'm wondering if this should be done prior to entering the first state handler.

bacongobbler replied:
As an aside, this comment would be great to add to the doc comments above on line 7.

kflansburg commented on crates/wasi-provider/src/states/running.rs:
```diff
@@ -73,17 +27,22 @@ impl State<PodState> for Running {
             {
                 if failed {
-                    return Transition::next(self, Error { message });
+                    // This appears to be required by the test `test_module_exiting_with_error`
```
I don't agree with `fail_fatal` here or in the init pod state, but it appears to be required by the integration tests. The tests expect the Pod to fail whereas I would expect it to go to `Error`. Incidentally there appears to be a bug when the Pod does go to `Error` and then back to `Registered`, the wasm runtime encounters some error with a channel. We may need to do some clean up in the error state in anticipation of going back to `Registered`.

bacongobbler replied:
Would it be all right if we pulled this out into a separate PR? That way we can have this discussion separate from the container status patching logic.

itowlson replied:
There was discussion about this on the 'remove panics from handlers' PR - this was to replicate existing behaviour which was relied on by the integration tests.  I think everyone agrees that it is not the desired behaviour, but changing the behaviour should be tackled as a specific work item (which will be non-trivial), not rolled in along with other work.

itowlson replied:
Never mind, I see you already raised an issue to do this as a separate item.  Thanks!

kflansburg commented on crates/wasi-provider/src/states/starting.rs:
```diff
@@ -109,14 +108,14 @@ impl State<PodState> for Starting {
+                Ok(h) => {
+                    container_handles.insert(ContainerKey::App(container.name().to_string()), h);
+                }
+                // We should log, transition to running, and properly handle container failure.
```
This is somewhat tied to making all of these handlers infallible, but this one in particular looked like it would drop important container status patches. The same can be said of the `Initializing` state handler.

kflansburg commented on crates/wasi-provider/src/wasi_runtime.rs:
```diff
@@ -299,9 +299,20 @@ impl WasiRuntime {
-                    return Err(anyhow::anyhow!(
-                    "_start import was not a function. This is likely a problem with the module"
-                ))
+                    let message = "_start import was not a function. This is likely a problem with the module";
```
This was the one Err in this method that did not notify the channel.

bacongobbler replied:
Good catch. I swore I updated this to the new notify logic...

kflansburg reviewed (commented):
This was a first crack at simplifying container status patches. There may be some nits here, but I think this is supporting work for the state machine, so I would expect it to get merged and then the container state machine will likely involve further refactoring.

bacongobbler commented:
Looks great. Going to take this for a spin to make sure things work as expected. But the fact that the container status patching logic was removed from the wasi provider is huge!

itowlson reviewed (changes requested):
This will be great to have!  Unfortunately I have reservations about relying so absolutely on the order of container statuses in the status section being in sync with the order of specs in the spec section.  It seems like if anything causes this assumption to become violated we have no way to detect and adapt (or even fail informatively) - we will just blindly write statuses to the wrong container.  Are we that confident about the assumption, and if not can we come up with an alternative approach?

kflansburg commented on crates/kubelet/src/container/status.rs:
```diff
@@ -82,3 +85,90 @@ impl Status {
+    pod: &Pod,
+    container_name: &str,
+    status: &Status,
+    init: bool,
```
Did not realize `ContainerKey` included init vs. app. Will change this.

kflansburg commented on crates/kubelet/src/container/status.rs:
```diff
@@ -82,3 +85,90 @@ impl Status {
+    match containers
+        .iter()
+        .enumerate()
+        .find(|(_, container)| container.name() == container_name)
```
Makes sense.

kflansburg commented on crates/kubelet/src/container/status.rs:
```diff
@@ -82,3 +85,90 @@ impl Status {
+        .find(|(_, container)| container.name() == container_name)
+    {
+        Some((container_index, container)) => {
+            let mut patches: Vec<json_patch::PatchOperation> = Vec::with_capacity(3);
```
There are four, I do not know if the resource version is actually necessary, but it was in the original WASI provider code and I added it back when I was debugging. In general I dont think its an issue to use a "magic number" in this case, as long as the number is actually correct. As far as the vec literal, that seems reasonable, although I will probably not create 3+ methods for that, just define the ReplaceOperation inside the vec literal.

kflansburg commented on crates/kubelet/src/container/status.rs:
```diff
@@ -82,3 +85,90 @@ impl Status {
+            ));
+
+            let patch = json_patch::Patch(patches);
+            let mut params = kube::api::PatchParams::default();
```
Yes, I believe that syntax works.

thomastaylor312 replied:
Yep, the syntax is

```rust
let thing = PatchParams {
    foo: "bar",
    ..Default::default()
};
```

kflansburg commented on crates/kubelet/src/container/status.rs:
```diff
@@ -82,3 +85,90 @@ impl Status {
+                format!("/status/initContainerStatuses/{}", container_index)
+            } else {
+                format!("/status/containerStatuses/{}", container_index)
+            };
```
I do not think it is guaranteed. I dont think it is reasonable to fetch the Pod each time it needs to be patched. We could store some up-to-date version of the Pod status, and overwrite the entire status when we patch a container. Is there any expectation that things other than than the scheduled kubelet update the status of the Pod under normal operation?

thomastaylor312 replied:
FWIW, I tried to write this as a JSON patch the first time I was tackling it as I didn't like the idea of fetching the pod each time. This ordering problem was the stickler then too. However, I think we might be able to remove some of the overhead of fetching the Pod each time if we switch to having a reflector. A reflector implementation updates its cache each time an object updates, so then we can just fetch the object from the cache. We could still end up with a case where that object updates after we fetch from the cache, but that should be fairly unlikely as I don't think anything else but the master could update a pod status. What I don't know is how many controllers might try to update pod statuses though.

kflansburg commented:
I have addressed all comments, except for switching to a Reflector. What I think makes the most sense is that `Pod::queue` should update the Pod definition that it is driving the state machine with whenever updates arrive. This would effectively be a reflector, and mean that the `Pod` passed to `json_status` will be the latest definition. I think this is the simplest solution and avoids having some shared state of Pod definitions that `patch_container_status` would need access to.

kflansburg commented on crates/kubelet/src/state.rs:
```diff
@@ -254,44 +257,49 @@ pub async fn run_to_completion<PodState: Send + Sync + 'static>(
+    let api: Api<KubePod> = Api::namespaced(client.clone(), &namespace);
+
+    // Initialize Pod Container Statuses
+    patch_status(&api, &name, make_registered_status(&initial_pod)).await;
```
I have implemented my suggestion of applying this initial status update prior to applying the Provider's first status. I also removed the use of this function in the Providers to verify that this eliminates their responsibility to do this.

kflansburg reviewed (commented):
This last commit implements my suggestion of updating the Pod manifest that is supplied to the State handlers.

thomastaylor312 commented on crates/kubelet/src/state.rs:
```diff
@@ -254,44 +257,49 @@ pub async fn run_to_completion<PodState: Send + Sync + 'static>(
+
+    // Initialize Pod Container Statuses
+    patch_status(&api, &name, make_registered_status(&initial_pod)).await;
+    drop(initial_pod);
```
Nit: consider doing this in a block (along with the `read()` call) so as to avoid the need for a manual `drop`

thomastaylor312 reviewed (approved):
I think with the new method of doing this, I am now ok with merging. Let's wait on @itowlson for his final review

thomastaylor312 commented on crates/kubelet/src/state.rs:
```diff
@@ -258,14 +266,62 @@ pub async fn run_to_completion<PodState: Send + Sync + 'static>(
+    // I'm not sure if we want to loop forever or handle some sort of failure
+    // condition (if Kubernetes refuses to accept and propagate this
+    // initialization patch.)
+    'main: loop {
```
Ok, think I have a few concerns about this. This is mainly broken down into two things 1) I don't know if this is any better than our old method of doing things and 2) If we do continue doing it this way, I think there could be a better way of doing this. 

1. So I don't know if this provides any benefit on our previous version other than API call reduction. We still have a bunch of waiting and things that we end up doing (no matter how we implement the retry). I think it could actually end up being slower as things step through the state machine
2. If we do want to go this route, I think we should reuse/export the `retry!` macros from the node code at the very minimum. I think a more elegant solution could entail wrapping the pod in a custom struct with a `RwLock` that write locks until it receives a response from the API and then abstract that all away behind `pod.read()` like we already have

kflansburg replied:
So regarding performance, I think this is better for two reasons:
* this loop only runs when a Pod is first scheduled, not on each state change.
* in most cases, all it does is wait a second or two for the reflector to update, and we don't have to try to patch multiple times. 
I will address the other nits.

thomastaylor312 replied:
> this loop only runs when a Pod is first scheduled, not on each state change.

Ok, so I misread that. I do think we should have a max number of retries (preferably as a `const`) like what you have now and then just error out and let k8s know the pod has errors (this might be a good use of the `Unknown` phase)

kflansburg replied:
Sounds good will do.

thomastaylor312 reviewed (changes requested):
I have some concerns with the loop approach in the latest commit. Detailed in the comments below

kflansburg commented on crates/kubelet/src/pod/mod.rs:
```diff
@@ -153,17 +154,26 @@ impl Pod {
+        let statuses: Vec<KubeContainerStatus> = if key.is_init() {
+            self.kube_pod
+                .status
+                .clone()
```
I dont think you can really do this: `.unwrap_or_else(|| &PodStatus::default())`, because the reference doesnt outlive the closure. I will see what I can do though.
